internal/kubernetes/drainSchedule.go

```
@@ -217,23 +220,39 @@ func (d *DrainSchedules) newSchedule(node *v1.Node, when time.Time) *schedule {
 	sched.timer = time.AfterFunc(time.Until(when), func() {
 		log := LoggerForNode(node, d.logger)
 		tags, _ := tag.New(context.Background(), tag.Upsert(TagNodeName, node.GetName())) // nolint:gosec
-		d.eventRecorder.Event(nr, core.EventTypeWarning, eventReasonDrainStarting, "Draining node")
-		if err := d.drainer.Drain(node); err != nil {
-			sched.finish = time.Now()
-			sched.setFailed()
-			log.Info("Failed to drain", zap.Error(err))
-			tags, _ = tag.New(tags, tag.Upsert(TagResult, tagResultFailed), tag.Upsert(TagFailureCause, string(getFailureCause(err)))) // nolint:gosec
-			StatRecordForEachCondition(tags, node, GetConditionsTypes(GetNodeOffendingConditions(node, d.suppliedConditions)), MeasureNodesDrained.M(1))
-			d.eventRecorder.Eventf(nr, core.EventTypeWarning, eventReasonDrainFailed, "Draining failed: %v", err)
+		replacementRequestEventDone := false // Flag to be sure that we produce the event only once.
+		if d.hasPreProvisioningAnnotation(node) {
+			log.Info("Start pre-provisioning before drain")
 			if err := RetryWithTimeout(
 				func() error {
-					return d.drainer.MarkDrain(node, when, sched.finish, true)
+					replacementStatus, err := d.drainer.PreProvisionNode(node, nodeReplacementReasonPreProvisioning)
+					if err != nil {
+						log.Error("Failed to validate node-replacement status", zap.Error(err))
+						return err
+					}
+					if !replacementRequestEventDone {
+						d.eventRecorder.Event(nr, core.EventTypeNormal, eventReasonNodePreProvisioning, "Node pre-provisioning before drain: request done.")
+						replacementRequestEventDone = true
+					}
+					if replacementStatus == NodeReplacementStatusDone {
+						d.eventRecorder.Event(nr, core.EventTypeNormal, eventReasonNodePreProvisioningCompleted, "Node pre-provisioning before drain: completed.")
+						return nil
+					}
+					return &NodePreProvisioningTimeoutError{} // Note that this error is dropped by RetryWithTimeout func.
```

I'm not a big fan of RetryWithTimeout. If I return an error, I expect it to be handled by the caller. RetryWithTimeout has several smelly side-effects:
- you log the error from PreProvisionNode AND return it 🙅 ;
- you use NodePreProvisioningTimeoutError as a dummy, which is confusing because the caller creates the same error when RetryWithTimeout gives up.

Why not user PollImmediate directly?

---

I agree, this function is sending the error to a black hole.
I think `RetryWithTimeout` was used instead of `PollImmediate` because the prototype of the function that we are passing defer by a boolean (the err is also playing the role of the boolean).
I am going to evaluate if we can surface the error for `RetryWithTimeout` without impacting too much the existing code/behavior.

---

It is not going to make it:
```
// ConditionFunc returns true if the condition is satisfied, or an error
// if the loop should be aborted.
```
In case of error we don't want to abord, we want to retry...

---

So I am going to keep `RetryWithTimeout` and explicitly return an error that we no is going to be dropped.

---

I tried `PolImmediate` but that makes the code a bit more complex due to the fact that I want to retry on error if the error is due to GET/UPDATE fail on the node inside PreProvision function.

---

> explicitly return an error that we no is going to be dropped

Should be a boolean.

As you know, for context, ConditionFunc is usually a wrapper function that handles retriable errors (e.g., log err) and returns false, nil; or throws non-retriable errors (i.e., returns false, err) -- in which case the loop is aborted but the caller needs to handle the error --; or returns true, nil on success. It throws its own timeout error if ConditionFunc never returns true, nil.

So RetryWithTimeout is like PollImmediate, when all errors are retriable, and not worth handling. That works when it's used with MarkDrain, but here, in a lambda, you do handle what PreProvisionNode returns, and at that point throwing an error is meaningless.

Maybe a RetryPredicateWithTimeout would make sense:

```
func RetryPredicateWithTimeout(f func() bool, retryPeriod, timeout time.Duration) error {
	return wait.PollImmediate(retryPeriod, timeout,
		func() (bool, error) {
			return f(), nil
		})
}
```